[Wong Wei Jian] iP

weijianwong opened the pull request:
# **Bobby App - Your Personal Assistant for Task Management**

> “The secret of getting ahead is getting started.” – Mark Twain

**Bobby** is your personal assistant for managing tasks efficiently and staying organized. It offers:

- A simple and intuitive interface
- Smart reminders and notifications
- Customizable task categories

### **How to Use Bobby:**

1. **Download** the Bobby app from [this link](https://github.com/weijianwong/ip).
2. **Open** the app and set up your profile.
3. **Create** and manage your tasks effortlessly.
4. **Stay organized** and let Bobby handle the rest! 🌟

#### **Features:**

- [ ] Task management with due dates
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Here’s a snippet from the core functionality:

```java
public static void main(String[] args) {
        new Bobby().run();
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

weijianwong committed:
Add Bobby.java

Add greet() which prints out the default greeting message.
Add exit() which prints out the default exit message.
Remove DUKE from the source code
Rename the file to Bobby.java

weijianwong committed:
Add feature to continuously prompt and echo user input

Implement a loop in the main method to prompt the user for input and echoes the input.
Add a condition to exit the loop when the user types "bye".

weijianwong committed:
Add list feature to the chatbot

Implement the ability to store user entered inputs to a fixed size array
Add a 'list' command to display all tasks in the order they were added.

weijianwong committed:
Add Task class and feature to mark and unmark tasks

Add tasks class to encapsulate functions and fields for tasks
Add feature to mark and unmark tasks based on the task number in the array

weijianwong committed:
Create new Task subtypes

Add feature to add different tasks including Deadline, Todo and Event.

weijianwong committed:
Add Automated text ui testing for Bobby

weijianwong committed:
Add custom exceptions to Bobby

Handle invalid list number when marking and unmarking.
Handle invalid todo, deadline and event inputs.

weijianwong committed:
Add delete feature to list

Add exceptions for delete feature.

weijianwong committed:
Add enums.

Add enums that include all the commands for Bobby.
Update the main function.

weijianwong committed:
Add feature to write inputs to a data file and read from the data file when program starts.

Previously, there is no feature to save inputs from users.

Having this feature makes it more user friendly as users do not have to manually input their tasks every time the program runs.

Save the data from the user input as [Task type] | [IsDone] | [Task Name] to make it easier to query the data. For Event and Deadline data, they have additional columns to store the deadline date and the event from and to data respectively.

weijianwong committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add feature to write inputs to a data file and read from the data file when program starts.

weijianwong committed:
Change dates from String to LocalDate

Add feature in Bobby to parse date string inputs and convert it to a LocalDate in each Task.

weijianwong committed:
Add new FIND command that finds for tasks based on the user input date.

weijianwong committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add new FIND command that finds for tasks based on the user input date.
  Change dates from String to LocalDate

weijianwong committed:
Create new classes to abstract out logic in Bobby class.

Ui class to handle all interactions with users.
Storage class to handle loading and saving of data to the Bobby.txt file.
Parser class to handle user inputs.
TaskList class to maintain the lists of tasks and provide method to add, remove and update tasks status.
Remove the echo function in Bobby as it is not used.

weijianwong committed:
Update Bobby class to use the other classes.

Abstract logic in Bobby class to other classes.
Update Bobby class to use the other classes. This makes the code easier to read and change in the future.

weijianwong committed:
Update Bobby class to use the other classes.

Added the data file to the gitignore.

weijianwong committed:
Merge branch 'A-MoreOOP'

* A-MoreOOP:
  Update Bobby class to use the other classes.
  Create new classes to abstract out logic in Bobby class.

weijianwong committed:
Add packages for bobby, command, exceptions, parser, storage, tasklist and tasks.

Previously all the files were stored under main/java. However, this was very messy as more classes are added.

Add packages to group related classes together to make the code neater.

weijianwong committed:
Merge branch 'A-Packages'

* A-Packages:
  Add packages for bobby, command, exceptions, parser, storage, tasklist and tasks.

weijianwong committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

weijianwong committed:
Add JUnit tests.

Unit testing can help to ensure that each function in each class is working correctly.

Lets add unit testing for the storage and parser class to test the methods inside.

weijianwong committed:
Merge branch 'A-JUnit'

* A-JUnit:
  Add JUnit tests.

weijianwong committed:
Package app as a jar file.

weijianwong committed:
Merge branch 'A-Jar'

* A-Jar:
  Package app as a jar file.

weijianwong committed:
Add javadoc header to all non private functions.

weijianwong committed:
Add javadoc headers to all classes

weijianwong committed:
Edit switch indentations to comply with coding standards

weijianwong committed:
Add function to find a task by date and keyword

Previously there was a function to find task by date only. To make the search more effective, we can add a function to find by keyword as well.

Add FINDDATE and FINDKEY commands to the Command enum.
Implement findTasksByDate and findTasksByKeyword methods in the Parser class.
Update parseFindCommand in Parser to handle both finddate and findkey commands.
Modify Bobby class to integrate handling of FINDDATE and FINDKEY commands and update the run method accordingly.

weijianwong committed:
Merge branch 'A-JavaDoc'

* A-JavaDoc:
  Add javadoc headers to all classes
  Add javadoc header to all non private functions.

weijianwong committed:
Merge branch 'A-CodingStandard'

* A-CodingStandard:
  Edit switch indentations to comply with coding standards

weijianwong committed:
Merge branch 'Level-9'

* Level-9:
  Add function to find a task by date and keyword

# Conflicts:
#	src/main/java/bobby/Bobby.java
#	src/main/java/bobby/parser/Parser.java
#	src/main/java/bobby/tasklist/TaskList.java
#	src/main/java/bobby/ui/Ui.java

weijianwong committed:
Add checkstyle to Bobby and modify files to comply with style checks

weijianwong committed:
Merge branch 'A-CheckStyle'

* A-CheckStyle:
  Add checkstyle to Bobby and modify files to comply with style checks

zi-yii commented on src/main/java/bobby/Bobby.java:
```diff
@@ -0,0 +1,117 @@
+                    ui.showTaskDeleted(deletedTask, tasks.size());
+                    storage.saveTasks(tasks);
+                    break;
+                case SEARCHDATE:
```
An explicit `//Fallthrough` comment should be included whenever there is a `case` statement without a break statement.

zi-yii commented on src/main/java/bobby/parser/Parser.java:
```diff
@@ -0,0 +1,136 @@
+package bobby.parser;
+
+import bobby.command.Command;
+import bobby.exceptions.*;
```
Wildcard imports are not allowed, imported classes should always be listed explicitly.

zi-yii commented on src/main/java/bobby/storage/Storage.java:
```diff
@@ -0,0 +1,125 @@
+     * with each task serialized to a specific format to facilitate easy loading later.
+     * If an I/O error occurs during the writing process, an error message is printed to the console.
+     *
+     * @param tasks the {@code TaskList} containing the tasks to be saved
```
Punctuation is required behind each parameter description.

zi-yii commented on src/test/java/bobby/parser/TestParser.java:
```diff
@@ -0,0 +1,88 @@
+package bobby.parser;
+
+import bobby.command.Command;
+import bobby.exceptions.*;
+import bobby.tasks.Task;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
```
I think static imports should be ordered first, followed by org.*, then the remaining imports.
Ordering of imports is shown [here](https://se-education.org/guides/tutorials/intellijCodeStyle.html).

zi-yii commented on src/main/java/bobby/storage/Storage.java:
```diff
@@ -0,0 +1,125 @@
+        }
+        return tasks;
+    }
+}
```
JavaDoc comments seems good!

zi-yii commented on src/main/java/bobby/Bobby.java:
```diff
@@ -0,0 +1,117 @@
+        new Bobby().run();
+    }
+
+}
```
I like that your code is quite clean 😃

zi-yii commented on src/main/java/bobby/parser/Parser.java:
```diff
@@ -0,0 +1,136 @@
+            throw new InvalidInputException();
+        }
+    }
+}
```
I find your JavaDoc comments very helpful here, well done 👍

zi-yii reviewed (commented):
Overall, I think you did a great job as your code is easy to read and your JavaDoc comments contributed to the understanding of your code 👍 
There are just a few nits to fix such as the ordering of imports and punctuations behind parameter description for JavaDoc comments.

weijianwong committed:
Add GUI to Bobby using JavaFX that echoes the user's input

weijianwong committed:
Edit Bobby class to return the correct response for each input

- Refactor getResponse method to integrate with the main application loop, providing actual command responses instead of echoing input.
- Refactor Ui class to improve code cleanliness and maintainability

weijianwong committed:
Merge branch 'Level-10'

* Level-10:
  Edit Bobby class to return the correct response for each input
  Add GUI to Bobby using JavaFX that echoes the user's input

weijianwong committed:
Add markMultipleTasks method to TaskList class to return processed tasks and add Javadoc

There is no method to mark multiple tasks at once in the TaskList class.

Allowing the user to mark multiple tasks be useful as the user does not need to key mark so many times is he wants to mark or unmark multiple tasks.

Create the markMultipleTasks method to return an ArrayList<Task> with the tasks that are marked or unmarked. Add Javadoc comments to describe the method’s parameters, return value, and exceptions.

Returning the processed tasks allows for better tracking of which tasks were affected which can be used by the UI class to display the information.

Error handling for NumberFormatException and IndexOutOfBoundsException is included, throwing InvalidInputException and InvalidTaskNumberException respectively. This ensures robust handling of invalid input and out-of-bounds indices.

weijianwong committed:
Merge branch 'A-Varargs'

* A-Varargs:
  Add markMultipleTasks method to TaskList class to return processed tasks and add Javadoc

weijianwong committed:
Add assertions for input validation to Parser and TaskList classes

The Parser class parses user input for commands and task creation. Input validation relies on exception handling but lacks explicit checks for some assumptions. Similarly, the TaskList class also lacks explicit checks for some assumptions like whether the keyword used in searches is null.

This needs to be changed to ensure that certain assumptions about input format and content are validated explicitly and to aid in debugging during development.

What is being done:
- Add assertions to parseFindCommand to verify that the command is either "searchdate" or "find".
- Add assertions to parseTask to ensure that the input format for "todo", "deadline", and "event" commands is correct and contains the expected number of elements.
- Add assertions to get, remove and add in TaskList to ensure that the index is not less than 0.
- Add assertions to findTasksByKeyword and findTasksByDate to ensure that keyword and date are not null.

Assertions help document and enforce assumptions about the input format during development. They are used in conjunction with exception handling to catch and diagnose potential issues more effectively.

weijianwong committed:
Refactor task parsing logic to adhere to SLAP principle

Current situation:
Splitting user input when marking and unmarking is currently done in the main Bobby class. This results in multiple levels of abstraction within the same method, decreasing readability and maintainability.

Why it needs to change:
To improve code quality and readability, it's essential to follow the Single Level of Abstraction Principle (SLAP). Keeping parsing logic separate from command execution makes the code easier to understand and maintain.

What is being done about it:
Moved task parsing logic to a dedicated method `extractTaskIndices` in the Parser class. This isolates the task parsing functionality, leaving the main Bobby class to focus only on command execution.

Why it is done that way:
This change ensures that each method operates at a single level of abstraction, enhancing code clarity and maintainability. Parsing-related logic now resides in the appropriate Parser class, improving separation of concerns.

weijianwong committed:
Refactor processCommand in Bobby to include default error handling in switch statement

Current situation:
The `processCommand` method in the Bobby application adds new tasks (Todo, Event, Deadline) in the default branch of the `switch` statement, which limits error detection for unrecognized commands.

Why it needs to be changed:
Using the `default` branch for task creation makes it difficult to handle unexpected commands and errors. It violates the best practice of using the `default` branch for catching unhandled cases, leading to potentially error-prone behavior.

What is being done about it:
Refactor the code to:
- Move task creation (Todo, Event, Deadline) to their own specific cases.
- Use the `default` branch to raise `InvalidInputException` for unknown commands.
- Explicitly handle all recognized commands, improving error detection.

Why is it done this way:
This approach ensures that all possible command outcomes are explicitly handled, following the Single Level of Abstraction Principle (SLAP) and improving the robustness and clarity of the `switch` statement.

weijianwong committed:
Refactor `saveTasks` Method to Use Streams for Improved Code Quality

Current situation:
The `saveTasks` method currently uses a traditional for-loop to iterate over tasks and write them to a file. This approach, while functional, results in more verbose and less readable code.

What is being done:
Refactor this method to use Java Streams. Specifically, a stream is created from the list of tasks, each task is converted to its string representation using `toFileString()`, and the results are collected and written to the file using `FileWriter`. This approach simplifies the code and leverages modern Java features for cleaner implementation.

Why is it done this way:
Using Streams improves code readability. It also aligns the code with current Java best practices, making it easier to maintain and extend in the future.

weijianwong committed:
Merge pull request #1 from weijianwong/branch-A-Assertions

Add assertions for input validation to Parser and TaskList classes

weijianwong committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions for input validation to Parser and TaskList classes

# Conflicts:
#	src/main/java/bobby/parser/Parser.java

weijianwong committed:
Merge branch 'master' into branch-A-Streams

* master:
  Add assertions for input validation to Parser and TaskList classes

weijianwong committed:
Merge pull request #2 from weijianwong/branch-A-CodeQuality

Pull request for branch A-code-quality

weijianwong committed:
Merge pull request #3 from weijianwong/branch-A-Streams

Refactor `saveTasks` Method to Use Streams for Improved Code Quality

weijianwong committed:
Merge branch 'master' of github.com:weijianwong/ip

* 'master' of github.com:weijianwong/ip:
  Refactor `saveTasks` Method to Use Streams for Improved Code Quality
  Refactor processCommand in Bobby to include default error handling in switch statement
  Refactor task parsing logic to adhere to SLAP principle

weijianwong committed:
Add gradle.yml file for continuous integration

weijianwong committed:
Merge branch 'A-CI'

* A-CI:
  Add gradle.yml file for continuous integration

weijianwong committed:
Remove unused imports in TaskList class and change assertions in Parser class

Previously assertions in Parser class are used wrongly to catch exceptions which led to build failure as test cases in the test file were failing.

This commit will delete the assertions in the Parser class as these exceptions are handled explicitly by the try catch block in the Parser class so there is no need for assert statements here.

weijianwong committed:
Enhance search functionality to support partial matches

Current situation:
- The search functionality only allows exact matches for task descriptions

What is being done:
- Implement partial matching in the `findTasksByKeyword` method to allow more flexible searches.

Why it's being done:
- To improve the flexibility and usability of the search feature by supporting partial keyword matches.

weijianwong committed:
Enhance Date Parsing and Search Functionality

Current Situation:
The parseDate method is not flexible enough to handle partial date inputs or specific keywords for dates like "today" and "tomorrow".

What is Being Done:
Update parseDate method to handle special keywords "today" and "tomorrow".

Why is it Done This Way:
Flexibility in Date Parsing: The parseDate method now supports both relative date keywords and absolute dates, providing more flexibility for date-related queries.

weijianwong committed:
Enhance the delete feature to remove multiple tasks at once

Add deleteMultipleTasks method to sort indices in descending order before deletion. This prevents issues caused by index shifting when tasks are removed from the list.

weijianwong committed:
Merge branch 'C-BetterSearch'

* C-BetterSearch:
  Enhance the delete feature to remove multiple tasks at once
  Enhance Date Parsing and Search Functionality
  Enhance search functionality to support partial matches

weijianwong committed:
Add duplicate task detection in TaskList class

weijianwong committed:
Add new InvalidTaskException to handle invalid tasks

weijianwong committed:
Modify add function in TaskList to check for duplicate tasks

weijianwong committed:
Add TestTaskList class to test the TaskList class

weijianwong committed:
Merge branch 'BCD-Extension'

* BCD-Extension:
  Add TestTaskList class to test the TaskList class
  Modify add function in TaskList to check for duplicate tasks
  Add new InvalidTaskException to handle invalid tasks
  Add duplicate task detection in TaskList class

weijianwong committed:
Refactor parser class to use command classes

weijianwong committed:
Update the runtest.bat and runtest.sh files

weijianwong committed:
Merge branch 'Refactor-Command'

* Refactor-Command:
  Update the runtest.bat and runtest.sh files
  Refactor parser class to use command classes

weijianwong committed:
Update test files for github actions

weijianwong committed:
Add installation for dos2unix to gradle.yml

weijianwong committed:
Remove the println in storage for testing

weijianwong committed:
Refactor the storage class to follow SLAP

weijianwong committed:
Change the javadoc header comments for the main method

weijianwong committed:
Refactor the saveTasks method to reduce arrow head code

weijianwong committed:
Merge branch 'Refactor-Storage'

* Refactor-Storage:
  Refactor the saveTasks method to reduce arrow head code
  Change the javadoc header comments for the main method
  Refactor the storage class to follow SLAP

weijianwong committed:
Modify the GUI to make it more user friendly

weijianwong committed:
Add javadoc header comments for javafx classes

weijianwong committed:
Merge branch 'A-BetterGui'

* A-BetterGui:
  Add javadoc header comments for javafx classes
  Modify the GUI to make it more user friendly

weijianwong committed:
Add welcome message to GUI

weijianwong committed:
Add Bobby title to java GUI

weijianwong committed:
Add Ui.png

weijianwong committed:
Edit readme in docs to explain all features of Bobby

weijianwong committed:
Merge branch 'A-UserGuide'

* A-UserGuide:
  Edit readme in docs to explain all features of Bobby

weijianwong committed:
Remove the top padding of dialogbox

weijianwong committed:
Modify the MainWindow to close after user enters bye

weijianwong committed:
Fix bugs with UI and event parser

weijianwong committed:
Add newline to DialogBox.java file

weijianwong committed:
Fix readme format
